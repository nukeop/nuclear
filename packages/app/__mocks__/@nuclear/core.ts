const initialStoreState = () => ({
  equalizer: {
    selected: 'Default'
  },
  downloads: [],
  favorites: {
    albums: [],
    tracks: []
  },
  playlists: []
});

let mockStore = initialStoreState();

module.exports = {
  store: {
    get: (key: string) => mockStore[key] || {},
    set: (key: string, value: any) => {
      mockStore[key] = value;
    },
    clear: () => mockStore = initialStoreState()
  },
  createApi: () => ({
    app: {},
    store: {},
    React: jest.fn(),
    ReactDOM: jest.fn()
  }),
  getOption: () => '',
  rest: {
    LastFmApi: class {
      constructor() {}

      getTagInfo() {}
      getTagTracks() {}
      getTagAlbums() {}
      getTagArtists() {}
      getTopTags = () => Promise.resolve()
      getTopTracks = () => Promise.resolve()
    }
  },
  settingsConfig: jest.requireActual('@nuclear/core/src/settings').settingsConfig,
  SettingType: {
    BOOLEAN: 'boolean',
    LIST: 'list',
    NODE: 'node',
    NUMBER: 'number',
    STRING: 'string',
    DIRECTORY: 'directory'
  }
};
