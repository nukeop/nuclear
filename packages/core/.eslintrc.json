{
  "extends": [
    "../../.eslintrc.json"
  ],
  "env": {
    "node": true,
    "browser": true,
    "commonjs": true
  },
  "rules": {
    "node/no-missing-import": 0,
    "node/no-extraneous-import": 0
  }
}
