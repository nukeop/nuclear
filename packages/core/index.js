export {
  LastFmApi,
  isElectron
} from './lib';
