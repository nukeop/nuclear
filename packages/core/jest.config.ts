export default async () => ({
  transform: {
    '^.+\\.(ts|tsx|js|jsx)$': 'ts-jest'
  },
  moduleNameMapper: {
    '^react$': '<rootDir>/../../node_modules/react'
  }
});
