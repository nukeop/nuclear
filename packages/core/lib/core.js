'use strict';

module.exports = core;

function core() {
  // TODO
}
