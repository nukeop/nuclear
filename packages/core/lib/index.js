import 'isomorphic-fetch';
export { default as LastFmApi } from './rest/lastfm';
