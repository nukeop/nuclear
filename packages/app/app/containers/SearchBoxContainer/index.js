import React from 'react';
import _ from 'lodash';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { compose, withHandlers, withProps } from 'recompose';
import { withTranslation } from 'react-i18next';
import { SearchBox } from '@nuclear/ui';

import * as SearchActions from '../../actions/search';
import * as PluginActions from '../../actions/plugins';

const MIN_SEARCH_LENGTH = 3;
const SearchBoxContainer = ({
  handleSearch,
  unifiedSearchStarted,
  searchProvidersOptions,
  selectedSearchProviderOption,
  handleSelectSearchProvider,
  isConnected,
  t,
  isFocused,
  handleFocus
}) => (
  <SearchBox
    loading={unifiedSearchStarted}
    disabled={!isConnected}
    placeholder={t('placeholder')}
    onChange={_.debounce(handleSearch, 500)}
    onSearch={handleSearch}
    searchProviders={searchProvidersOptions}
    selectedSearchProvider={selectedSearchProviderOption}
    onSearchProviderSelect={handleSelectSearchProvider}
    isFocused={isFocused}
    handleFocus={handleFocus}
  />
);

const mapStateToProps = state => ({
  unifiedSearchStarted: state.search.unifiedSearchStarted,
  isConnected: state.connectivity,
  searchProviders: state.plugin.plugins.metaProviders,
  selectedSearchProvider: state.plugin.selected.metaProviders,
  isFocused: state.search.isFocused
});

const mapDispatchToProps = dispatch => ({
  searchActions: bindActionCreators(SearchActions, dispatch),
  pluginActions: bindActionCreators(PluginActions, dispatch)
});

const providerToOption = provider => ({
  key: _.lowerCase(_.get(provider, 'sourceName')),
  text: _.get(provider, 'searchName'),
  value: _.get(provider, 'sourceName')
});

export default compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps),
  withTranslation('search'),
  withHandlers({
    handleSearch:
      ({ searchActions, history }) => value =>
        value.length >= MIN_SEARCH_LENGTH ? searchActions.unifiedSearch(value, history) : null,
    handleSelectSearchProvider:
      ({ pluginActions }) => provider =>
        pluginActions.selectMetaProvider(provider.value),
    handleFocus: ({ searchActions }) => bool => searchActions.setSearchDropdownVisibility(bool)
  }),
  withProps(({ searchProviders, selectedSearchProvider }) => ({
    searchProvidersOptions: _.map(searchProviders, providerToOption),
    selectedSearchProviderOption: providerToOption(
      _.find(searchProviders, { sourceName: selectedSearchProvider })
    )
  }))
)(SearchBoxContainer);
