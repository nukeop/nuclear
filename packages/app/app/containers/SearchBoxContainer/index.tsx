import React, { useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { MetaProvider } from '@nuclear/core';
import { SearchBox } from '@nuclear/ui';
import { SearchActions, unifiedSearch } from '../../actions/search';
import { pluginsSelectors } from '../../selectors/plugins';
import { searchSelectors } from '../../selectors/search';
import { connectivity } from '../../selectors/connectivity';
import { selectMetaProvider } from '../../actions/plugins';

const MIN_SEARCH_LENGTH = 2;

const providerToOption = (provider?: MetaProvider) => provider && ({
  key: provider.sourceName.toLowerCase(),
  text: provider.searchName,
  value: provider.sourceName
});

const SearchBoxContainer: React.FC = () => {
  const { t } = useTranslation('search');
  const history = useHistory();
  const dispatch = useDispatch();
  const unifiedSearchStarted = useSelector(searchSelectors.unifiedSearchStarted);
  const isConnected = useSelector(connectivity);
  const searchProviders = useSelector(pluginsSelectors.plugins).metaProviders as MetaProvider[];
  const searchHistory = useSelector(searchSelectors.searchHistory);
  const selectedSearchProvider = useSelector(pluginsSelectors.selected)?.metaProviders;
  const isFocused = useSelector(searchSelectors.isFocused);
  const handleFocus = useCallback((focusState: boolean) => dispatch(SearchActions.setSearchDropdownVisibility(focusState)), [dispatch]);

  const searchProvidersOptions = searchProviders?.map(providerToOption) ?? [];
  const selectedSearchProviderOption = providerToOption(
    searchProviders?.find(provider => provider.sourceName === selectedSearchProvider)
  );

  const handleSearch = useCallback((value: string) => {
    if (value.length >= MIN_SEARCH_LENGTH) {
      handleFocus(false);
      return dispatch(unifiedSearch(value, history));
    }
    return null;
  },
  [dispatch, history]
  );
  const handleClearSearchHistory = useCallback(() => dispatch(SearchActions.updateSearchHistory([])), [dispatch]);
  const handleSelectSearchProvider = useCallback((provider: ReturnType<typeof providerToOption>) =>
    provider && dispatch(selectMetaProvider(provider.value)),
  [dispatch]);


  return <SearchBox 
    loading={unifiedSearchStarted}
    disabled={!isConnected}
    placeholder={t('placeholder')}
    lastSearchesLabel={t('last-searches')}
    clearHistoryLabel={t('clear-history')}
    footerLabel={t('you-can-search-for')}
    onSearch={handleSearch}
    searchProviders={searchProvidersOptions}
    searchHistory={searchHistory}
    onClearHistory={handleClearSearchHistory}
    selectedSearchProvider={selectedSearchProviderOption}
    onSearchProviderSelect={handleSelectSearchProvider}
    isFocused={isFocused}
    handleFocus={handleFocus}
  />;
};

export default SearchBoxContainer;
