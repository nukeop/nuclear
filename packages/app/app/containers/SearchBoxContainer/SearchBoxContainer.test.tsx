import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { rest } from '@nuclear/core';

import SearchBoxContainer from '.';
import { AnyProps, configureMockStore, setupI18Next, TestStoreProvider } from '../../../test/testUtils';
import { buildStoreState } from '../../../test/storeBuilders';


describe('Search box container', () => {
  beforeAll(() => {
    setupI18Next();
    jest.useFakeTimers();
  });

  it('should render the search box', () => {
    const { component } = mountComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should search after pressing enter', () => {
    const { component, store } = mountComponent();
    const searchInput = component.getByTestId('search-input');
    fireEvent.change(searchInput, { target: { value: 'test' } });
    fireEvent.keyDown(searchInput, { key: 'Enter', keyCode: 13 });

    const state = store.getState();
    expect(state.search.searchHistory).toEqual(['test']);
    expect(state.plugin.plugins.metaProviders[0].searchForArtists).toHaveBeenCalledWith('test');
    expect(state.plugin.plugins.metaProviders[0].searchForReleases).toHaveBeenCalledWith('test');
    expect(state.plugin.plugins.metaProviders[0].searchForPodcast).toHaveBeenCalledWith('test');

    const lastfmApi = new rest.LastFmApi('test', 'test');
    expect(lastfmApi.searchTracks).toHaveBeenCalledWith('test');
    expect(rest.Youtube.urlSearch).toHaveBeenCalledWith('test');
    expect(rest.Youtube.liveStreamSearch).toHaveBeenCalledWith('test');
  });

  it('should select the metadata provider from the dropdown', () => {
    const { component, store } = mountComponent();
    component.getAllByText('Test Metadata Provider')[0].click();
    component.getByText('Another Metadata Provider').click();

    const state = store.getState();
    expect(state.plugin.selected.metaProviders).toEqual('Another Metadata Provider');
  });

  const mountComponent = (initialStore?: AnyProps) => {
    const initialState = initialStore ||
            buildStoreState()
              .withPlugins()
              .build();
    const store = configureMockStore(initialState);
    const component = render(
      <TestStoreProvider store={store}>
        <SearchBoxContainer />
      </TestStoreProvider>
    );

    return {
      component,
      store
    };
  };
});
