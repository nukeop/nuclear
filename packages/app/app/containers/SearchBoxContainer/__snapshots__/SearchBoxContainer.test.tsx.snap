// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search box container should render the search box 1`] = `
<DocumentFragment>
  <div
    class="nuclear search_box_container"
  >
    <div
      class="nuclear search_box"
    >
      <i
        aria-hidden="true"
        class="search icon"
      />
      <input
        class=""
        data-testid="search-input"
        placeholder="Search..."
        value=""
      />
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ui dropdown"
        role="listbox"
        tabindex="0"
      >
        <div
          aria-atomic="true"
          aria-live="polite"
          class="divider text"
          role="alert"
        >
          Test Metadata Provider
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <div
            aria-checked="true"
            aria-selected="true"
            class="active selected item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Test Metadata Provider
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <span
              class="text"
            >
              Another Metadata Provider
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
